File: shared/src/commonMain/kotlin/com/loapalette/shared/LorcanaCard.kt
package com.loapalette.shared

import kotlin.experimental.ExperimentalObjCName
import kotlin.native.ObjCName
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class LorcanaCardApiResponse(
        @SerialName("Name") val name: String? = null,
        @SerialName("Cost") val cost: Int? = null,
        @SerialName("Type") val type: String? = null,
        @SerialName("Color") val color: String? = null,
        @SerialName("Inkable") val inkable: Boolean? = null,
        @SerialName("Body_Text") val bodyText: String? = null,
        @SerialName("Flavor_Text") val flavorText: String? = null,
        @SerialName("Artist") val artist: String? = null,
        @SerialName("Rarity") val rarity: String? = null,
        @SerialName("Image") val image: String? = null,
        @SerialName("Set_Name") val setName: String? = null,
        @SerialName("Set_ID") val setId: String? = null,
        @SerialName("Set_Num") val setNum: Int? = null,
        @SerialName("Strength") val strength: Int? = null,
        @SerialName("Willpower") val willpower: Int? = null,
        @SerialName("Lore") val lore: Int? = null,
        @SerialName("Classifications") val classifications: String? = null
) {
    fun toLorcanaCard(): LorcanaCard {
        val abilitiesList =
                bodyText?.let {
                    if (it.isNotBlank()) {
                        listOf(it)
                    } else {
                        null
                    }
                }

        val classificationsList =
                classifications?.split(",")?.map { it.trim() }?.filter { it.isNotBlank() }

        return LorcanaCard(
                id = setId?.let { "$it-${setNum ?: ""}" },
                name = name,
                cost = cost,
                color = color,
                inkwell = inkable,
                type = type,
                rarity = rarity,
                set = setId,
                setNumber = setNum,
                flavorText = flavorText ?: "",
                illustrator = artist,
                imageUrl = image,
                abilities = abilitiesList,
                strength = strength,
                willpower = willpower,
                lore = lore
        )
    }
}

// 既存のLorcanaCardモデル（互換性のため維持）
@OptIn(ExperimentalObjCName::class)
@ObjCName("LorcanaCard", exact = true)
@Serializable
data class LorcanaCard(
        val id: String? = null,
        val name: String? = null,
        val cost: Int? = null,
        val color: String? = null,
        val inkwell: Boolean? = null,
        val type: String? = null,
        val rarity: String? = null,
        val set: String? = null,
        val setNumber: Int? = null,
        val flavorText: String? = null,
        val illustrator: String? = null,
        val imageUrl: String? = null,
        val abilities: List<String>? = null,
        val strength: Int? = null,
        val willpower: Int? = null,
        val lore: Int? = null
)

// カードバリアント
@Serializable
data class CardVariant(
        val set: String? = null,
        val id: Int? = null,
        val dreamborn: String? = null,
        val ravensburger: RavensburgerInfo? = null,
        val rarity: String? = null,
        val illustrator: String? = null
)

// Ravensburger情報
@Serializable
data class RavensburgerInfo(
        val en: String? = null,
        val fr: String? = null,
        val de: String? = null,
        val it: String? = null,
        val zh: String? = null,
        val ja: String? = null,
        @SerialName("culture_invariant_id") val cultureInvariantId: Int? = null,
        @SerialName("sort_number") val sortNumber: Int? = null
)

@Serializable
data class CardLanguages(
        val en: LanguageData? = null,
        val fr: LanguageData? = null,
        val de: LanguageData? = null,
        val zh: LanguageData? = null,
        val ja: LanguageData? = null
)

// 言語データ
@Serializable
data class LanguageData(val name: String, val title: String, val flavour: String? = null)

@Serializable
data class CardClassification(val slug: String, val en: String, val fr: String? = null)

@Serializable
data class CardAbility(
        val type: String? = null,
        val title: AbilityTitle? = null,
        val text: AbilityText? = null,
        val ability: String? = null
)

@Serializable data class AbilityTitle(val en: String? = null)

@Serializable data class AbilityText(val en: String? = null)

@Serializable data class ImageUris(val digital: DigitalImageUris? = null)

@Serializable
data class DigitalImageUris(
        val small: String? = null,
        val normal: String? = null,
        val large: String? = null
)

@Serializable
data class CardsResponse(
        val cards: List<LorcanaCard> = emptyList(),
        val total: Int? = null,
        val page: Int? = null,
        val pageSize: Int? = null
)

public object ImageUrlGenerator {
    private const val BASE_URL = "https://lorcana-api.com/images"
    private const val APOSTROPHE = "\u0027"

    private val nameShortenMap = mapOf("tramp" to "tram")

    private fun sanitizeForUrl(text: String): String {
        return text.lowercase()
                .replace("\u2019", APOSTROPHE)
                .replace("\u2018", APOSTROPHE)
                .replace("\u0060", APOSTROPHE)
                .replace("\u02BC", APOSTROPHE)
                .replace("\u02BB", APOSTROPHE)
                .replace("\u02C8", APOSTROPHE)
                .replace(",", "")
    }

    private fun toSnakeCase(text: String): String {
        return sanitizeForUrl(text).replace(" ", "_").replace("-", "_")
    }

    private fun titleToUrlPath(title: String): String {
        return sanitizeForUrl(title).replace(" ", "_")
    }

    fun generateFallbackImageUrl(enLanguage: LanguageData?): String? {
        val enName = enLanguage?.name ?: return null

        val rawTitle = enLanguage.title.trim()
        val enTitle = if (rawTitle.isBlank()) "" else rawTitle

        var characterName = toSnakeCase(enName)
        characterName = nameShortenMap[characterName] ?: characterName

        val fileName = toSnakeCase(enName)

        return if (enTitle.isEmpty()) {
            "$BASE_URL/$characterName/$fileName-large.png"
        } else {
            val cardTitle = titleToUrlPath(enTitle)
            val fileTitle = titleToUrlPath(enTitle)
            "$BASE_URL/$characterName/$cardTitle/$fileName-$fileTitle-large.png"
        }
    }
}

File: iosApp/LoaPalette/LoaPalette/Views/CardSearch/Models/LorcanaCard.swift
//
//  LorcanaCard.swift
//  LoaPalette
//
//  Created by 片岡寿哉 on 2025/11/28.
//

import Foundation
import shared

// 画像URI群（Lorcast API）
struct ImageUris: Codable {
    let digital: DigitalImageUris?
}

// デジタル画像URI群
struct DigitalImageUris: Codable {
    let small: String?
    let normal: String?
    let large: String?
}

// カードデータモデル
// 参考: https://api-lorcana.com/#/Cards/get%20cards
struct LorcanaCard: Codable, Identifiable {
    let cardId: String?
    let name: String?
    let cost: Int?
    let color: String?
    let inkwell: Bool?
    let type: String?
    let rarity: String?
    let set: String?
    let setNumber: Int?
    let flavorText: String?
    let illustrator: String?
    let imageUrl: String?
    let abilities: String?  // カンマ区切りの文字列
    let strength: Int?
    let willpower: Int?
    let lore: Int?

    // 新しいAPI構造のフィールド
    let attack: Int?
    let defence: Int?
    let variants: [CardVariant]?
    let languages: CardLanguages?
    let classifications: [CardClassification]?
    let abilitiesArray: [CardAbility]?
    let imageUris: ImageUris?

    enum CodingKeys: String, CodingKey {
        case cost
        case inkwell
        case color
        case type
        case lore
        case attack
        case defence
        case variants
        case languages
        case classifications
        case abilitiesArray = "abilities"
        case illustrator
        case imageUris = "image_uris"
    }

    // カスタムデコード処理
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        cost = try container.decodeIfPresent(Int.self, forKey: .cost)
        inkwell = try container.decodeIfPresent(Bool.self, forKey: .inkwell)
        color = try container.decodeIfPresent(String.self, forKey: .color)
        type = try container.decodeIfPresent(String.self, forKey: .type)
        lore = try container.decodeIfPresent(Int.self, forKey: .lore)
        attack = try container.decodeIfPresent(Int.self, forKey: .attack)
        defence = try container.decodeIfPresent(Int.self, forKey: .defence)
        variants = try container.decodeIfPresent([CardVariant].self, forKey: .variants)
        languages = try container.decodeIfPresent(CardLanguages.self, forKey: .languages)
        classifications = try container.decodeIfPresent(
            [CardClassification].self, forKey: .classifications)
        abilitiesArray = try container.decodeIfPresent([CardAbility].self, forKey: .abilitiesArray)
        illustrator = try container.decodeIfPresent(String.self, forKey: .illustrator)
        imageUris = try container.decodeIfPresent(ImageUris.self, forKey: .imageUris)

        // 日本語データを優先的に使用
        if let jaData = languages?.ja {
            name = jaData.name
            flavorText = jaData.flavour ?? ""
        } else if let enData = languages?.en {
            name = enData.name
            flavorText = enData.flavour ?? ""
        } else {
            name = nil
            flavorText = nil
        }

        // バリアントから情報を取得（最初のバリアントを使用）
        if let firstVariant = variants?.first,
            let variantId = firstVariant.id
        {
            cardId = String(variantId)
            // レアリティを大文字始まりに変換（Common, Uncommon, Rare, Super Rare, Legendary）
            if let variantRarity = firstVariant.rarity {
                let rarityMap: [String: String] = [
                    "common": "Common",
                    "uncommon": "Uncommon",
                    "rare": "Rare",
                    "super_rare": "Super Rare",
                    "legendary": "Legendary",
                ]
                rarity =
                    rarityMap[variantRarity] ?? variantRarity.prefix(1).uppercased()
                    + variantRarity.dropFirst()
            } else {
                rarity = nil
            }
            set = firstVariant.set?.uppercased()
            // セット番号を取得（例: "1/204 JA 1" から "1" を取得）
            if let jaString = firstVariant.ravensburger?.ja,
                let firstPart = jaString.split(separator: "/").first,
                let number = Int(firstPart)
            {
                setNumber = number
            } else {
                setNumber = nil
            }
        } else {
            cardId = nil
            rarity = nil
            set = nil
            setNumber = nil
        }

        // 画像URLを構築（Lorcast APIのimage_uris.digital.largeを優先、なければフォールバック）
        // 優先順位: image_uris.digital.large > 手動構築URL
        if let largeImageUrl = imageUris?.digital?.large {
            // Lorcast APIから提供される画像URLを使用
            imageUrl = largeImageUrl
        } else if let en = languages?.en {
            // Kotlin側の共通実装を使用
            let kotlinLanguageData = shared.LanguageData(
                name: en.name,
                title: en.title,
                flavour: en.flavour
            )
            imageUrl = shared.ImageUrlGenerator.shared.generateFallbackImageUrl(
                enLanguage: kotlinLanguageData)
        } else {
            #if DEBUG
                print("⚠️ Missing name for image URL generation")
            #endif
            imageUrl = nil
        }

        // 攻撃力・防御力のマッピング
        strength = attack
        willpower = defence

        // アビリティを文字列に変換
        if let abilitiesArray = abilitiesArray, !abilitiesArray.isEmpty {
            abilities = abilitiesArray.compactMap {
                $0.title?.en ?? $0.ability ?? $0.type
            }.filter { !$0.isEmpty }.joined(separator: ", ")
        } else {
            abilities = nil
        }
    }

    // エンコード処理（必要に応じて実装）
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(cost, forKey: .cost)
        try container.encodeIfPresent(inkwell, forKey: .inkwell)
        try container.encodeIfPresent(color, forKey: .color)
        try container.encodeIfPresent(type, forKey: .type)
        try container.encodeIfPresent(lore, forKey: .lore)
        try container.encodeIfPresent(attack, forKey: .attack)
        try container.encodeIfPresent(defence, forKey: .defence)
        try container.encodeIfPresent(variants, forKey: .variants)
        try container.encodeIfPresent(languages, forKey: .languages)
        try container.encodeIfPresent(classifications, forKey: .classifications)
        try container.encodeIfPresent(abilitiesArray, forKey: .abilitiesArray)
        try container.encodeIfPresent(illustrator, forKey: .illustrator)
    }

    // Identifiableプロトコルに必要なidプロパティ
    var id: String {
        return cardId ?? name ?? UUID().uuidString
    }
}

// カードバリアント
struct CardVariant: Codable {
    let set: String?
    let id: Int?
    let dreamborn: String?
    let ravensburger: RavensburgerInfo?
    let rarity: String?
    let illustrator: String?
}

// Ravensburger情報
struct RavensburgerInfo: Codable {
    let en: String?
    let fr: String?
    let de: String?
    let it: String?
    let zh: String?
    let ja: String?
    let cultureInvariantId: Int?
    let sortNumber: Int?

    enum CodingKeys: String, CodingKey {
        case en, fr, de, it, zh, ja
        case cultureInvariantId = "culture_invariant_id"
        case sortNumber = "sort_number"
    }
}

// カード言語情報
struct CardLanguages: Codable {
    let en: LanguageData?
    let fr: LanguageData?
    let de: LanguageData?
    let zh: LanguageData?
    let ja: LanguageData?
}

// 言語データ
struct LanguageData: Codable {
    let name: String
    let title: String
    let flavour: String?
}

// カード分類
struct CardClassification: Codable {
    let slug: String
    let en: String
    let fr: String?
}

// カードアビリティ
struct CardAbility: Codable {
    let type: String?
    let title: AbilityTitle?
    let text: AbilityText?
    let ability: String?
}

// アビリティタイトル
struct AbilityTitle: Codable {
    let en: String?
}

// アビリティテキスト
struct AbilityText: Codable {
    let en: String?
}

// APIエラーレスポンス
struct APIError: Codable {
    let code: String
    let details: String
    let object: String
    let status: Int
}

// SearchState enum
enum SearchState: String {
    case idle = "IDLE"
    case loading = "LOADING"
    case success = "SUCCESS"
    case error = "ERROR"
}
